documentation/architecture.md

```diff
@@ -0,0 +1,336 @@
+
+Docker's `containerd-shim` is designed around the assumption that it can monitor and reap the actual
+container process. As `containerd-shim` runs on the host, it can not directly monitor a process running
+within a virtual machine. At most it can see the QEMU process but that is not enough.
+With Clear Containers, `cc-shim` acts as the container process that `containerd-shim` can monitor. Therefore
+`cc-shim` needs to handle all container I/O streams (`stdout`, `stdin` and `stderr`) and forward all signals
+`containerd-shim` decides to send to the container process.
+
+`cc-shim` has an implicit knowledge about which VM agent will handle those streams and signals and thus act as
+an encapsulation layer between `containerd-shim` and `hyperstart`:
+
+- It fragments and encapsulates the standard input stream from containerd-shim into `hyperstart` stream packets:
+```
+  ┌─────────────────┬───────────────┬────────────────────┐
+  │    IO stream    │ Packet length │ IO stream fragment │
+  │ sequence number │    (32 bits)  │                    │
+  └─────────────────┴───────────────┴────────────────────┘
+```
+- It deencapsulates and assemble standard output and error `hyperstart` stream packets into an output stream
+that it forwards to `containerd-shim`
+- It translates all UNIX signals (except `SIGKILL` and `SIGSTOP`) into `hyperstart` ``KILLCONTAINER` commands
+that it sends to the VM via `cc-proxy` UNIX named socket.
+
+The IO stream sequence numbers are passed from `cc-proxy` to `cc-shim` when the former spawns the latter.
+They are generated by `hyperstart` and `cc-oci-runtime` fetches them by sending the `AllocateIO` command to
+`cc-proxy`.
+
+As an example, let's say that running the `# pwd` command from a container standard input will generate
+`/tmp` from the container standard output. `hyperstart` assigned this specific process 8888 and 8889 respectively
+as the stdin, stdour and stderr sequence numbers.
+With `cc-shim` and Clear Containers, this example would look like:
+
+![cc-shim](https://github.com/01org/cc-oci-runtime/blob/topic/documentation/documentation/shim.png)
+
+#### Networking
+
+Containers will typically live in their own, possibly shared, networking namespace.
+At some point in a container lifecycle, container engines will set up that namespace
+to add the container to a network which is isolated from the host network, but which is shared between containers
+
+In order to do so, container engines will usually add one end of a `virtual ethernet
+(veth)` pair into the container networking namespace. The other end of the `veth` pair
+is added to the container network.
+
+This is a very namespace-centric approach as QEMU can not handle `veth` interfaces.
+Instead it typically creates `TAP` interfaces for adding connectivity to a virtual
+machine.
+
+To overcome that incompatibility between typical container engines expectations
+and virtual machines, `cc-oci-runtime` networking implementation intents to transparently
```

Comment:
'intents'. Not sure that is the right word here - but struggling to find a suitable one word replacement. Hmm, maybe 'implements' ?